gocourse02/main.go

```diff
@@ -0,0 +1,127 @@
+package main
+
+import "fmt"
+
+type Animal struct {
+	name    string
+	species string
+	cage    *Cage
+}
+
+type Zoo struct {
+	Animals []*Animal
+}
+
+type Cage struct {
+	name   string
+	animal *Animal
+}
+
+type AllCages struct {
+	Cages []Cage
+}
+
+type Zookeeper struct {
+	name string
+}
+
+func NewAnimal(species string, name string) *Animal {
+	return &Animal{species: species, name: name}
+}
+func (a *Animal) Escape() {
+	fmt.Printf("Animal %s escaped from the cage %v\n", a.name, a.cage.name)
+	a.cage.animal = nil
+	a.cage = nil
+}
+func NewCage(name string) *Cage {
+	return &Cage{name: name}
+}
+
+func NewZookeeper() *Zookeeper {
+	return &Zookeeper{name: "Bob"}
+}
+
+func (c Cage) Describe() {
+	if c.animal == nil {
+		fmt.Println("Cage is empty!")
+	} else {
+		fmt.Printf("Cage %s is occupied by animal: %s\n", c.name, c.animal.name)
+	}
+}
+
+func (a Animal) Reproduce(species string, name string) *Animal {
+	return NewAnimal(species, name)
+}
+
+// AddAnimalToCage садить всіх звірів із зоопарку в клітки.
+// Повертає nil, якщо тварина поміщена в клітку.
+// Повертає помилку, якщо всі клітки зайняті і тварина не поміщена в клітку.
+func (zk Zookeeper) AddAnimalToCage(c *AllCages, z *Zoo) error {
```

Comment:
Пропоную видалити структуру `AllCages`, вона не сильно корисна та ускладнює код. Перенести `Cages []Cage` в структуру `Zoo`. Ніби логічно, що зоопарк містить в собі звірів та клітки. 

Буде так:
```suggestion
func (zk Zookeeper) AddAnimalToCage(z *Zoo) error {
```

Можливо, навіть зоопарк містить в собі наглядача. Але це вже другорядне.